LibraBFT/Concrete/Properties/PreferredRound.agda

```diff
@@ -24,24 +25,216 @@ open import LibraBFT.Yasm.Yasm ℓ-RoundManager ℓ-VSFP ConcSysParms PeerCanSig
+                     → (vfr : VoteForRound∈ pk round epoch bId (msgPool st))
+                     → ¬ (∈GenInfo (ver-signature (msgSigned vfr)))
+                     → (sig : WithVerSig pk v)
+                     → v ^∙ vRound ≡ round
+                     → ¬ (∈GenInfo (ver-signature sig))
+   ¬Gen∧Round≡⇒¬Gen r pkH v₁ ¬genV₁ sigV₂ refl genV₂
+      with ¬genVotesRound≢0 r pkH (msgSigned v₁) (msg⊆ v₁) (msg∈pool v₁) ¬genV₁
+   ...| v₁r≢0 = ⊥-elim (v₁r≢0 (trans (msgRound≡ v₁) (genVotesRound≡0 sigV₂ genV₂)))
+
+   α-ValidVote-trans : ∀ {mbr vabs} (v₁ v₂ : Vote)
+                     → α-ValidVote 𝓔 v₁ mbr ≡ vabs
+                     → v₂ ^∙ vRound ≡ v₁ ^∙ vRound
+                     → v₂ ^∙ vEpoch ≡ v₁ ^∙ vEpoch
+                     → v₂ ^∙ vProposedId ≡ v₁ ^∙ vProposedId
+                     → α-ValidVote 𝓔 v₂ mbr ≡ vabs
+   α-ValidVote-trans v₁ v₂ refl refl refl refl = refl
+
+   PreferredRoundProof :
+      ∀ {pk round₁ round₂ epoch bId₁ bId₂ v₁abs v₂abs mbr} {st : SystemState}
+      → ReachableSystemState st
+      → Meta-Honest-PK pk
+      → (v₁ : VoteForRound∈ pk round₁ epoch bId₁ (msgPool st))
+      → (v₂ : VoteForRound∈ pk round₂ epoch bId₂ (msgPool st))
+      → round₁ < round₂
+      → α-ValidVote 𝓔 (msgVote v₁) mbr ≡ v₁abs
+      → α-ValidVote 𝓔 (msgVote v₂) mbr ≡ v₂abs
+      → (c3 : Cand-3-chain-vote v₁abs)
+      → Σ (VoteParentData v₂abs)
+            (λ vp → Cand-3-chain-head-round c3 ≤ Abs.round (vpParent vp))
+   PreferredRoundProof step@(step-s r theStep) pkH v₁ v₂ r₁<r₂ refl refl c3
+      with msgRound≡ v₁ | msgEpoch≡ v₁ | msgBId≡ v₁
+         | msgRound≡ v₂ | msgEpoch≡ v₂ | msgBId≡ v₂
+   ...| refl | refl | refl | refl | refl | refl
+      with ∈GenInfo? (₋vSignature (msgVote v₁)) | ∈GenInfo? (₋vSignature (msgVote v₂))
+   ...| yes init₁  | yes init₂  = let r₁≡0 = genVotesRound≡0 (msgSigned v₁) init₁
+                                      r₂≡0 = genVotesRound≡0 (msgSigned v₂) init₂
+                                  in ⊥-elim (<⇒≢ r₁<r₂ (trans r₁≡0 (sym r₂≡0)))
+   ...| yes init₁  | no  ¬init₂ = let 0≡rv = sym (genVotesRound≡0 (msgSigned v₁) init₁)
+                                      0<rv = v-cand-3-chain⇒0<roundv c3
+                                  in ⊥-elim (<⇒≢ 0<rv 0≡rv)
+   ...| no  ¬init₁ | yes init₂  = let 0≡r₂ = sym (genVotesRound≡0 (msgSigned v₂) init₂)
+                                      r₁   = msgVote v₁ ^∙ vRound
+                                  in ⊥-elim (<⇒≱ r₁<r₂ (subst (r₁ ≥_) 0≡r₂ z≤n))
+   ...| no  ¬init₁ | no ¬init₂
+      with theStep
+   ...| step-peer cheat@(step-cheat c)
+        = let m₁sb4 = ¬cheatForgeNewSig r cheat unit pkH (msgSigned v₁) (msg⊆ v₁) (msg∈pool v₁) ¬init₁
+              m₂sb4 = ¬cheatForgeNewSig r cheat unit pkH (msgSigned v₂) (msg⊆ v₂) (msg∈pool v₂) ¬init₂
+              v₁sb4 = msgSentB4⇒VoteRound∈ (msgSigned v₁) m₁sb4
+              v₂sb4 = msgSentB4⇒VoteRound∈ (msgSigned v₂) m₂sb4
+              v₁abs = α-ValidVote-trans (msgVote v₁) (msgVote v₁sb4) refl
```

Comment:
These four lines are repeated verbatim twice below, so it would be good to do some refactoring to reduce the redundancy.  One idea is to make a version of `α-ValidVote-trans` that takes a `VoteForRound∈` whose `msgVote` is the second `Vote`, rather than passing all the equalities, which are taken from such a `VoteForRound∈` anyway.  That makes the redundancy easier on the eye, but doesn't eliminate it.  Another possibility is to separate the two `with` abstractions at lines 184-187, so we first make the `v₁abs` in the `inj₂ v₁sb4` so we don't repeat it for both branches of the other one.

Reply:
You are right! Fixed!